object/init_data.go

```diff
@@ -106,7 +106,19 @@ func initDefinedApplication(application *Application) {
 		return
 	}
 	application.CreatedTime = util.GetCurrentTime()
-	AddApplication(application)
+	// If user not defined clientId and clientSecret
+	if len(application.ClientId) == 0 || len(application.ClientSecret) == 0 {
+		application.ClientId = util.GenerateClientId()
+		application.ClientSecret = util.GenerateClientSecret()
+	}
+	for _, providerItem := range application.Providers {
+		providerItem.Provider = nil
```

Comment:
Why set to `nil` exlicitly, isn't it `nil` by default?

Reply:
It is the same as `AddApplication`
https://github.com/casdoor/casdoor/blob/ec296215471b1e2e35add3826f07ec5bd8165970/object/application.go#L282-L295